jigglypuffRL/deeprl/sac/sac.py

```diff
@@ -0,0 +1,388 @@
+    :param update_interval: (int) Number of steps between parameter updates
+    :param save_interval: (int) Number of steps between saves of models
+    :param layers: (tuple or list) Number of neurons in hidden layers
+    :param tensorboard_log: (str) the log location for tensorboard (if None,
+        no logging)
+    :param seed (int): seed for torch and gym
+    :param render (boolean): if environment is to be rendered
+    :param device (str): device to use for tensor operations; 'cpu' for cpu
+        and 'cuda' for gpu
+        and 'cuda' for gpu
+    :param pretrained: (boolean) if model has already been trained
+    :param save_name: (str) model save name (if None, model hasn't been
+        pretrained)
+    :param save_version: (int) model save version (if None, model hasn't been
+        pretrained)
+    """
+
+    def __init__(
+        self,
+        network_type,
+        env,
+        gamma=0.99,
+        replay_size=1000000,
+        batch_size=256,
+        lr=3e-4,
+        alpha=0.01,
+        polyak=0.995,
+        entropy_tuning=True,
+        epochs=1000,
+        start_steps=0,
+        steps_per_epoch=1000,
+        max_ep_len=1000,
+        start_update=256,
+        update_interval=1,
+        save_interval=5000,
+        layers=(256, 256),
+        tensorboard_log=None,
+        seed=None,
+        render=False,
+        device="cpu",
+        pretrained=False,
+        save_name=None,
+        save_version=None,
+    ):
+
+        self.network_type = network_type
+        self.env = env
+        self.gamma = gamma
+        self.replay_size = replay_size
+        self.batch_size = batch_size
+        self.lr = lr
+        self.alpha = alpha
+        self.polyak = polyak
+        self.entropy_tuning = entropy_tuning
+        self.epochs = epochs
+        self.start_steps = start_steps
+        self.steps_per_epoch = steps_per_epoch
+        self.max_ep_len = max_ep_len
+        self.start_update = start_update
+        self.update_interval = update_interval
+        self.save_interval = save_interval
+        self.layers = layers
+        self.tensorboard_log = tensorboard_log
+        self.seed = seed
+        self.render = render
+        self.save_name = save_name
+        self.save_version = save_version
+        self.save = save_params
+        self.load = load_params
+        self.evaluate = evaluate
+        self.checkpoint = self.__dict__
+        self.pretrained = pretrained
+
+        # Assign device
+        if "cuda" in device and torch.cuda.is_available():
+            self.device = torch.device(device)
+        else:
+            self.device = torch.device("cpu")
+
+        # Assign seed
+        if seed is not None:
+            torch.manual_seed(seed)
+            torch.backends.cudnn.deterministic = True
+            torch.backends.cudnn.benchmark = False
+            np.random.seed(seed)
+            self.env.seed(seed)
+            random.seed(seed)
```

Comment:
Should we make a function to set these seeds? Pretty much being copied around in every algo

Reply:
Yeah we definitely should for this, tensorboard, device and training loop

Reply:
But shall we do it once all the algos are done so that it doesn't affect other algos being added

Reply:
Sounds good